File: .cursor/hooks/notion_daily_precheck.py
"""Daily precheck hook for Notion MCP calls.

Behavior:
1) Runs on beforeMCPExecution.
2) Only applies to Notion MCP server.
3) Executes once per day (first Notion task).
4) If precheck fails, triggers one-click repair.
"""

from __future__ import annotations

import json
import os
import platform
import socket
import subprocess
import sys
from datetime import datetime
from pathlib import Path
from typing import Any

from notion_auth_oneclick_fix import run_fix

STATE_PATH = Path("Daily-Backups") / "TMP_notion_precheck_state.json"
LOG_PATH = Path("Daily-Backups") / "TMP_notion_precheck_log.jsonl"
NOTION_SERVER_KEYWORD = "notion"


def _emit(payload: dict[str, Any]) -> None:
    sys.stdout.write(json.dumps(payload, ensure_ascii=False) + "\n")
    sys.stdout.flush()


def _read_stdin_payload() -> dict[str, Any]:
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        return json.loads(raw)
    except Exception:
        return {"raw_input": raw, "parse_error": True}


def _project_root() -> Path:
    return Path(os.environ.get("CURSOR_PROJECT_DIR") or os.getcwd())


def _safe_read_json(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {}
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return {}


def _write_json(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")


def _append_log(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as f:
        f.write(json.dumps(payload, ensure_ascii=False) + "\n")


def _is_notion_mcp(payload: dict[str, Any]) -> bool:
    text = json.dumps(payload, ensure_ascii=False).lower()
    return NOTION_SERVER_KEYWORD in text


def _extract_tool_name(payload: dict[str, Any]) -> str:
    for key in ("toolName", "tool_name", "mcp_tool_name", "name"):
        value = payload.get(key)
        if isinstance(value, str) and value:
            return value
    return ""


def _already_checked_today(state: dict[str, Any]) -> bool:
    today = datetime.now().strftime("%Y-%m-%d")
    return state.get("date") == today and state.get("ok") is True


def _check_dns() -> tuple[bool, str]:
    try:
        socket.gethostbyname("www.notion.so")
        return True, "dns_ok"
    except Exception as exc:  # noqa: BLE001
        return False, f"dns_failed:{exc}"


def _check_https_association() -> tuple[bool, str]:
    if platform.system().lower() != "windows":
        return True, "skip_non_windows"
    try:
        proc = subprocess.run(
            [
                "reg",
                "query",
                r"HKCU\Software\Microsoft\Windows\Shell\Associations\UrlAssociations\https\UserChoice",
                "/v",
                "ProgId",
            ],
            capture_output=True,
            text=True,
            timeout=6,
            check=False,
        )
        if proc.returncode != 0:
            return False, f"https_assoc_query_failed:{proc.returncode}"
        if "ProgId" not in proc.stdout:
            return False, "https_assoc_missing_progid"
        return True, "https_assoc_ok"
    except Exception as exc:  # noqa: BLE001
        return False, f"https_assoc_failed:{exc}"


def _check_interactive_session() -> tuple[bool, str]:
    session = os.environ.get("SESSIONNAME", "")
    if not session:
        return True, "session_unknown"
    if "service" in session.lower():
        return False, f"non_interactive_session:{session}"
    return True, f"interactive_session:{session}"


def _run_precheck() -> dict[str, Any]:
    checks = []
    dns_ok, dns_msg = _check_dns()
    checks.append({"name": "dns_resolve_notion", "ok": dns_ok, "detail": dns_msg})
    assoc_ok, assoc_msg = _check_https_association()
    checks.append({"name": "https_default_browser_association", "ok": assoc_ok, "detail": assoc_msg})
    session_ok, session_msg = _check_interactive_session()
    checks.append({"name": "interactive_desktop_session", "ok": session_ok, "detail": session_msg})
    return {"ok": all(c["ok"] for c in checks), "checks": checks}


def main() -> None:
    payload = _read_stdin_payload()
    if not _is_notion_mcp(payload):
        _emit({"permission": "allow"})
        return

    tool_name = _extract_tool_name(payload)
    if tool_name == "mcp_auth":
        _emit({"permission": "allow"})
        return

    root = _project_root()
    state_file = root / STATE_PATH
    log_file = root / LOG_PATH
    state = _safe_read_json(state_file)

    if _already_checked_today(state):
        _emit({"permission": "allow"})
        return

    precheck = _run_precheck()
    now = datetime.now().isoformat()
    log_entry: dict[str, Any] = {
        "time": now,
        "kind": "notion_daily_precheck",
        "ok": precheck["ok"],
        "tool_name": tool_name,
        "checks": precheck["checks"],
    }

    if precheck["ok"]:
        _write_json(
            state_file,
            {
                "date": datetime.now().strftime("%Y-%m-%d"),
                "ok": True,
                "updated_at": now,
            },
        )
        _append_log(log_file, log_entry)
        _emit({"permission": "allow"})
        return

    fix_result = run_fix()
    log_entry["fix"] = fix_result
    _write_json(
        state_file,
        {
            "date": datetime.now().strftime("%Y-%m-%d"),
            "ok": False,
            "updated_at": now,
            "last_error": precheck["checks"],
            "fix": fix_result,
        },
    )
    _append_log(log_file, log_entry)
    _emit(
        {
            "permission": "ask",
            "user_message": (
                "Notion 每日预检失败，已自动触发一键修复（浏览器唤起预热）。"
                "请确认当前网络/默认浏览器后，先执行一次 mcp_auth 再继续。"
            ),
            "agent_message": "Notion daily precheck failed and auto-repair was executed.",
        }
    )


if __name__ == "__main__":
    main()

File: adapters/cursor/hooks/notion_daily_precheck.mjs
/**
 * Daily precheck hook for Notion MCP calls.
 * Runs once per day on first Notion MCP task.
 */

import { readFileSync, writeFileSync, appendFileSync, mkdirSync, existsSync, writeSync } from "node:fs";
import { join } from "node:path";
import { lookup } from "node:dns/promises";
import { platform } from "node:os";
import { spawnSync } from "node:child_process";
import { runFix } from "./notion_auth_oneclick_fix.mjs";

const STATE_PATH = join("Daily-Backups", "TMP_notion_precheck_state.json");
const LOG_PATH = join("Daily-Backups", "TMP_notion_precheck_log.jsonl");

function emit(payload) {
  writeSync(1, Buffer.from(`${JSON.stringify(payload)}\n`, "utf8"));
}

function readPayload() {
  const raw = readFileSync(0, "utf8");
  if (!raw.trim()) return {};
  try {
    return JSON.parse(raw);
  } catch {
    return { raw_input: raw, parse_error: true };
  }
}

function projectRoot() {
  return process.env.CURSOR_PROJECT_DIR || process.cwd();
}

function readJson(filePath) {
  if (!existsSync(filePath)) return {};
  try {
    return JSON.parse(readFileSync(filePath, "utf8"));
  } catch {
    return {};
  }
}

function writeJson(filePath, payload) {
  mkdirSync(join(projectRoot(), "Daily-Backups"), { recursive: true });
  writeFileSync(filePath, JSON.stringify(payload, null, 2), "utf8");
}

function appendLog(filePath, payload) {
  mkdirSync(join(projectRoot(), "Daily-Backups"), { recursive: true });
  appendFileSync(filePath, `${JSON.stringify(payload)}\n`, "utf8");
}

function isNotionMcp(payload) {
  return JSON.stringify(payload).toLowerCase().includes("notion");
}

function extractToolName(payload) {
  return payload.toolName || payload.tool_name || payload.mcp_tool_name || payload.name || "";
}

function todayDate() {
  return new Date().toISOString().slice(0, 10);
}

function alreadyCheckedToday(state) {
  return state.date === todayDate() && state.ok === true;
}

async function checkDns() {
  try {
    await lookup("www.notion.so");
    return { name: "dns_resolve_notion", ok: true, detail: "dns_ok" };
  } catch (err) {
    return { name: "dns_resolve_notion", ok: false, detail: `dns_failed:${String(err)}` };
  }
}

function checkHttpsAssociation() {
  if (platform() !== "win32") {
    return { name: "https_default_browser_association", ok: true, detail: "skip_non_windows" };
  }
  try {
    const proc = spawnSync(
      "reg",
      [
        "query",
        "HKCU\\Software\\Microsoft\\Windows\\Shell\\Associations\\UrlAssociations\\https\\UserChoice",
        "/v",
        "ProgId",
      ],
      { encoding: "utf8", timeout: 6000 }
    );
    if ((proc.status ?? 1) !== 0) {
      return {
        name: "https_default_browser_association",
        ok: false,
        detail: `https_assoc_query_failed:${proc.status}`,
      };
    }
    if (!String(proc.stdout || "").includes("ProgId")) {
      return {
        name: "https_default_browser_association",
        ok: false,
        detail: "https_assoc_missing_progid",
      };
    }
    return { name: "https_default_browser_association", ok: true, detail: "https_assoc_ok" };
  } catch (err) {
    return {
      name: "https_default_browser_association",
      ok: false,
      detail: `https_assoc_failed:${String(err)}`,
    };
  }
}

function checkInteractiveSession() {
  const session = process.env.SESSIONNAME || "";
  if (!session) {
    return { name: "interactive_desktop_session", ok: true, detail: "session_unknown" };
  }
  if (session.toLowerCase().includes("service")) {
    return { name: "interactive_desktop_session", ok: false, detail: `non_interactive_session:${session}` };
  }
  return { name: "interactive_desktop_session", ok: true, detail: `interactive_session:${session}` };
}

async function runPrecheck() {
  const checks = [await checkDns(), checkHttpsAssociation(), checkInteractiveSession()];
  return { ok: checks.every((c) => c.ok), checks };
}

async function main() {
  const payload = readPayload();
  if (!isNotionMcp(payload)) {
    emit({ permission: "allow" });
    return;
  }

  const toolName = extractToolName(payload);
  if (toolName === "mcp_auth") {
    emit({ permission: "allow" });
    return;
  }

  const root = projectRoot();
  const stateFile = join(root, STATE_PATH);
  const logFile = join(root, LOG_PATH);
  const state = readJson(stateFile);
  if (alreadyCheckedToday(state)) {
    emit({ permission: "allow" });
    return;
  }

  const precheck = await runPrecheck();
  const now = new Date().toISOString();
  const logEntry = {
    time: now,
    kind: "notion_daily_precheck",
    ok: precheck.ok,
    tool_name: toolName,
    checks: precheck.checks,
  };

  if (precheck.ok) {
    writeJson(stateFile, { date: todayDate(), ok: true, updated_at: now });
    appendLog(logFile, logEntry);
    emit({ permission: "allow" });
    return;
  }

  const fix = runFix();
  logEntry.fix = fix;
  writeJson(stateFile, {
    date: todayDate(),
    ok: false,
    updated_at: now,
    last_error: precheck.checks,
    fix,
  });
  appendLog(logFile, logEntry);
  emit({
    permission: "ask",
    user_message:
      "Notion 每日预检失败，已自动触发一键修复（浏览器唤起预热）。请确认网络与默认浏览器后，先执行一次 mcp_auth。",
    agent_message: "Notion daily precheck failed and auto-repair was executed.",
  });
}

main();
